src/coredump/coredump-ratelimit.c

```diff
@@ -0,0 +1,167 @@
+        const char *p;
+        int i;
+
+        /* check if the file is a core dump file */
+        p = startswith(file_name, "core.");
+        if (!p) {
+                /* check if the file is an inprogress core dump file */
+                p = startswith(file_name, ".#core.");
+                if (!p)
+                        return -EINVAL;
+        }
+
+        /* Skip the comm field */
+        p = strchr(p, '.');
+        if (!p)
+                return -EINVAL;
+        p++;
+
+        /* Skip the uid field */
+        p = strchr(p, '.');
+        if (!p)
+                return -EINVAL;
+        p++;
+
+        /* Retrieve the Bootid */
+        for (i = 0; *p != '.' && i < SD_ID128_STRING_MAX - 1; p++, i++){
+                bootid[i] = *p;
+        }
+        bootid[i] = '\0';
+        return 0;
+}
+
+int coredump_ratelimit(const char* process_name, usec_t interval, unsigned burst) {
+
+        _cleanup_closedir_ DIR *d = NULL;
+        struct dirent *de;
+        unsigned int count = 0;
+        usec_t curr_ts;
+        sd_id128_t boot_id;
+        char sid[33];
+        int r;
+
+        /* If the process name is not given, then return -EINVAL. No rate limit happens.*/
+        if (process_name == NULL)
+                return -EINVAL;
+
+        /* either if burst or interval is zero then ratelimit is disabled */
+        if (burst == 0 || interval == 0)
+                return 0;  /* No ratelimit */
+
+        /* Whenever a daemon crashes, we compare the number of core dump files present in the
+         * /var/lib/systemd/coredump belonging to that daemon and generated within the
+         * rate limit interval with the rate limit. If the limit is reached, the core dump
+         * will not be processed further for generation.
+         */
+
+        d = opendir("/var/lib/systemd/coredump");
+        if (!d) {
+                if (errno == ENOENT)
+                        return -ENOENT;
+                return log_error_errno(errno, "Can't open coredump directory: %m");
+        }
+
+        /* get the current timestamp */
+        curr_ts = now(CLOCK_REALTIME);
+        log_debug("Current time="USEC_FMT" usecs", curr_ts);
+        log_debug("Ratelimit interval="USEC_FMT" usecs", interval);
+        /* Known limitation :
+         * If system time gets changed (advanced or delayed), then it results in a new
+         * rate limit window allowing the generation of new core dumps until the limit
+         * is reached again.
+         */
+
+        r = sd_id128_get_boot(&boot_id);
+        if (r < 0)
+                return r;
+        sd_id128_to_string(boot_id, sid);
+        log_debug("Bootid=%s\n", sid);
+        FOREACH_DIRENT_ALL(de, d, goto fail) {
+                char comm[TASK_COMM_LEN + 1];
+                char f_boot_id[33];
+                struct stat st;
+                usec_t ts;
+
+                r = comm_from_file_name(de->d_name, comm);
```

Comment:
i think "comm" is not a suitable identifier actually. it's just too short. i mean, look at systemd itself, it ships various binaries names systemd-journalxyz and they all have the same comm. 

I'd probably use either the exe name (i.e. the executable path), or maybe even the combination of the exe name and the comm name. That would improve beaviour for script engines, that do patch comm: we'd respect that and not consider all python programs as the same faulty process...

Reply:
Also, I'd much prefer if we'd use data from xattrs here, instead of parsing too much out of the name. we already attach exe + comm as xattrs to the files, it's trivial to read them off the files (use getxattr_malloc()). 

In fact, given that you apparently also want to filter by boot Id, maybe we should attach that to the coredump too, i.e. add a line or two to the end of fix_xattr() in coredump.c. that attach the boot id. 

That way, you can really easily determine the bits you need via getxattr_malloc() in a safe and reliable way without making the precise way we name coredumps API.

BTW, please also update the comment in fix_xattr(). It shouldn't say anymore "Just because we can" then, but say "Used by the ratelimiting logic".

Reply:
I have addressed the review comment. Please check once

Reply:
Current version is still not using xattr as suggested

Reply:
@Sribharath27 ping

Reply:
No problem and no rush, just ensuring it doesn't get lost. Get well soon!

Reply:
Apologies for the late response. Addressed other comments except for xattr. Could you please share some references? 
It would great help